File: web/src/core/planner.ts
import type { MediaInfo, Segment, SplitRequest } from "./domain";

export class SegmentPlanner {
  static readonly EPSILON = 1e-6;

  static nearestKeyframe(timestamp: number, keyframes: readonly number[]): number {
    if (keyframes.length === 0) return timestamp;
    return keyframes.reduce((closest, candidate) =>
      Math.abs(candidate - timestamp) < Math.abs(closest - timestamp)
        ? candidate
        : closest
    );
  }

  plan(request: SplitRequest, media: MediaInfo): readonly Segment[] {
    const target = request.segmentDuration;
    if (!Number.isFinite(target) || target <= 0) {
      throw new Error("Segment duration must be greater than zero.");
    }
    if (!Number.isFinite(media.duration) || media.duration <= 0) {
      throw new Error("Video duration must be greater than zero.");
    }

    const boundaries = [0];
    for (let nominal = target; nominal < media.duration; nominal += target) {
      let boundary = SegmentPlanner.nearestKeyframe(nominal, media.keyframes);
      boundary = clamp(boundary + request.offset, 0, media.duration);
      const previous = boundaries.at(-1) ?? 0;
      if (boundary <= previous + SegmentPlanner.EPSILON) {
        boundary = clamp(nominal + request.offset, 0, media.duration);
      }
      if (
        boundary > previous + SegmentPlanner.EPSILON &&
        boundary < media.duration
      ) {
        boundaries.push(boundary);
      }
    }
    boundaries.push(media.duration);

    const spans: Array<[number, number]> = boundaries.slice(0, -1).map((start, index) => [
      start,
      boundaries[index + 1] ?? media.duration
    ]);
    const finalSpan = spans.at(-1);
    if (finalSpan) {
      const [start, end] = finalSpan;
      const length = end - start;
      if (
        length > target &&
        length <= target * 1.1 &&
        !request.allowLongLastPart
      ) {
        spans.splice(spans.length - 1, 1, [start, start + target], [start + target, end]);
      }
    }

    const stem = safeStem(request.fileName);
    return spans.map(([start, end], zeroBasedIndex) => {
      const index = zeroBasedIndex + 1;
      const isFinal = index === spans.length;
      const duration = end - start;
      return {
        index,
        start,
        end,
        padDuration: isFinal && duration < target ? target - duration : 0,
        outputName: `${stem}-part${index}.mp4`
      };
    });
  }
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.min(Math.max(value, minimum), maximum);
}

function safeStem(fileName: string): string {
  const withoutExtension = fileName.replace(/\.[^/.]+$/, "");
  const sanitized = withoutExtension
    .normalize("NFKD")
    .replace(/[^a-zA-Z0-9._-]+/g, "-")
    .replace(/^-+|-+$/g, "");
  return sanitized || "video";
}

File: instagram_story_parts/planner.py
"""Pure, deterministic segment planning."""

from __future__ import annotations

from collections.abc import Callable, Iterable

from .domain import MediaInfo, Segment, SplitRequest


LongLastPartPolicy = Callable[[float], bool]


class SegmentPlanner:
    """Create a gap-free plan, using keyframes for internal boundaries."""

    _EPSILON = 1e-6

    @staticmethod
    def nearest_keyframe(timestamp: float, keyframes: Iterable[float]) -> float:
        candidates = tuple(keyframes)
        if not candidates:
            return timestamp
        return min(candidates, key=lambda value: abs(value - timestamp))

    def plan(
        self,
        request: SplitRequest,
        media: MediaInfo,
        allow_long_last_part: LongLastPartPolicy | None = None,
    ) -> tuple[Segment, ...]:
        target = request.segment_duration
        if target <= 0:
            raise ValueError("Segment duration must be greater than zero.")
        if media.duration <= 0:
            raise ValueError("Video duration must be greater than zero.")

        boundaries = [0.0]
        nominal = target
        while nominal < media.duration:
            boundary = self.nearest_keyframe(nominal, media.keyframes)
            boundary = min(max(boundary + request.offset, 0.0), media.duration)

            # A repeated/extreme keyframe must not collapse two planned parts.
            if boundary <= boundaries[-1] + self._EPSILON:
                boundary = min(max(nominal + request.offset, 0.0), media.duration)
            if boundary > boundaries[-1] + self._EPSILON and boundary < media.duration:
                boundaries.append(boundary)
            nominal += target
        boundaries.append(media.duration)

        spans = list(zip(boundaries, boundaries[1:]))
        last_start, last_end = spans[-1]
        last_length = last_end - last_start
        if target < last_length <= target * 1.1:
            keep_long = bool(
                allow_long_last_part and allow_long_last_part(last_length)
            )
            if not keep_long:
                spans[-1:] = [(last_start, last_start + target),
                              (last_start + target, last_end)]

        segments: list[Segment] = []
        output_dir = request.resolved_output_dir
        for index, (start, end) in enumerate(spans, start=1):
            pad_duration = 0.0
            if index == len(spans) and end - start < target:
                pad_duration = target - (end - start)
            segments.append(
                Segment(
                    index=index,
                    start=start,
                    end=end,
                    output_path=output_dir
                    / f"{request.video_path.stem}-part{index}.mp4",
                    pad_duration=pad_duration,
                )
            )
        return tuple(segments)
